### ISSUE TITLE
Invertible Encoder/Decoder?

### ISSUE THREAD

**Issue by rpellumbi** [VLL-HD/FrEIA#32]
We are trying to create a convolutional neural net such that the input is [batch_size, 259, 64, 64] and would like to output a size of [batch_size, 3, 64, 64]. Is there a way to build an encoder such that this is possible, and invertible, with the FrEIA architecture? 

Non specific to this example, does there exist an invertible encoder/decoder that results in the input and output having different shapes with a different amount of parameters.

**Comment by bionicles** [comment]
One way to achieve this would be to have placeholders involved (random noise) and use differently sized placeholder for different sized outputs ... but I only just discovered FrEIA and am not sure how to do that with this (yet)

**Comment by psteinb** [comment]
Invertible networks rely on a concept called normalizing flows. There the restriction is that your input data is of the same shape as the output. So what @bionicles mentioned is right on the spot. What you may want to do is: Have your input of shape `batch_size, 259, 64, 64` and the output have the same shape (which FrEIA does automatically in the OutputNode IIRC). But to start with, you want to fill the missing values of the output with random noise. 

This seminal paper by the VLL-HD group yields a description on how to do that:
https://arxiv.org/abs/1808.04730
See also the affiliated repo: https://github.com/VLL-HD/analyzing_inverse_problems

**Comment by ardizzone** [comment]
It might also be worth using a cINN, which we found much more stable and reliable for image data specifically.
Here, the invertibility is only between z <--> x, and y is given as an external condition.

We have this workshop paper comparing different architectures for low-dimesional problems:
https://invertibleworkshop.github.io/INNF_2019/accepted_papers/pdfs/INNF_2019_paper_6.pdf
It seems to work comparably, but gives me a lot less trouble getting to
work properly, usually works out of the box, and there are no worries
about dimensionality.

For images, you will either have to adapt the 'Analyzing-Inverse-Problems-INN' to also use max. likelihood training, or switch to the cINN.

Methodological publications about the cINN:

https://drive.google.com/drive/folders/1cX2B55B6iDf1zwCASQ-ezmDyCEjNQZWL?usp=sharing
(to appear in GCPR proceedings 2020)
Preprint version (some different experiments):
https://arxiv.org/abs/1907.02392

https://arxiv.org/pdf/2003.06281.pdf
https://bpspsychub.onlinelibrary.wiley.com/doi/abs/10.1111/bmsp.12159
https://github.com/stefanradev93/BayesFlow

There are several publications from other (partly collaborating) groups,
for which the cINN worked well for various science problems, low- and
high-dimensional:
https://academic.oup.com/mnras/article-abstract/499/4/5447/5950348
https://www.scipost.org/SciPostPhys.9.5.074/pdf
https://arxiv.org/abs/2012.08195
https://arxiv.org/abs/2011.05110
(I think there is one more from medicine, it should be somewhere on google scholar)

I will close this issue for now, but feel free to ask if you have any further questions!
